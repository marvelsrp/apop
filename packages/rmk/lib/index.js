const reduceObject = (origin, remakers) => {
  return remakers.reduce((state, remake) => {
    return remake(state);
  }, origin);
};

const reduceArray = (origin, remakers) => {
  return origin.reduce((state, element) => {
    state.push(reduceObject(element, remakers));
    return state;
  }, []);
};

const reduce = (origin, reamakers) => {
  if (isArray(origin)) {
    return reduceArray(origin, mutators);
  } else if (isObject(origin)) {
    return reduceObject(origin, reamakers);
  } else {
    throw new Error('Wrong input arguments');
  }
};

const rmk = (...reamakers) => {
  return origin => reduce(origin, reamakers);
};

export default rmk;
