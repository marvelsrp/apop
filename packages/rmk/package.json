{
  "name": "@rmk/rmk",
  "version": "1.0.0",
  "scripts": {
    "prepublish": "babel -d ./ ./lib",
    "dev": "babel -d ./ ./lib --watch",
    "test": "jest --coverage",
    "test-dev": "jest --watch"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tuchk4/rmk.git"
  },
  "dependencies": {
    "@rmk/clear": "=1.0.0"
  },
  "license": "ISC",
  "main": "index.js"
}
