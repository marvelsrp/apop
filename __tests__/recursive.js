import rmk from 'rmk';
import rmkActions from 'rmk/actions';
import mock from '../__mock__/recusrive';
import isArray from 'lodash/isArray';

describe('recursive', () => {

  it('object recursive', () => {
    let formula = rmk(
      rmkActions.recursive(
        [
          rmkActions.toCamelCase(),
          rmkActions.update({
            arrayKeyStr: (localState) =>  (isArray(localState.arrayKey)) ?  localState.arrayKey.join(',') : null
          }),
          rmkActions.rename({
            arrayKeyStr: 'renamedStr'
          }),
          rmkActions.clear()
        ]
      )
    );
    let result = formula(mock.recursiveBefore);
    expect(result).toEqual(mock.recursiveAfter);
  });
});