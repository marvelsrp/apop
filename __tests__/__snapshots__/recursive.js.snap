// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array recursive 1`] = `
Array [
  Object {
    "isCategory": false,
    "isLink": true,
    "name": "1 old",
    "url": "/1",
    "version": "v1",
  },
  Object {
    "isCategory": false,
    "isLink": true,
    "name": "2",
    "url": "/2",
    "version": "v2",
  },
  Object {
    "isCategory": true,
    "isLink": false,
    "isOpen": false,
    "nodes": Array [
      Object {
        "isCategory": false,
        "isLink": true,
        "name": "3",
        "url": "/3",
        "version": "v2",
      },
      Object {
        "isCategory": true,
        "isLink": false,
        "isOpen": false,
        "name": "6",
        "nodes": Array [
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "3 old",
            "url": "/3",
            "version": "v1",
          },
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "4",
            "url": "/4",
            "version": "v2",
          },
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "5",
            "url": "/5",
            "version": "v2",
          },
        ],
      },
    ],
  },
]
`;

exports[`Object recursive 1`] = `
Object {
  "arrayKey": Array [
    1,
    2,
    3,
  ],
  "arrayObjects": Array [
    Object {
      "aB": "a",
      "bC": "b",
    },
  ],
  "deepObj": Object {
    "arrayKey": Array [
      1,
      2,
      3,
    ],
    "arrayObjects": Array [
      null,
    ],
    "fooBar": 0,
    "renamedStr": "1,2,3",
  },
  "fooBar": 0,
  "renamedStr": "1,2,3",
}
`;
